be/src/olap/rowset/segment_v2/binary_dict_page.h

```diff
@@ -116,6 +116,10 @@ class BinaryDictPageDecoder : public PageDecoder {
         return _data_page_decoder->current_index();
     }
 
+    bool is_dict_encoding();
+
+    void set_dict_decoder(std::shared_ptr<BinaryPlainPageDecoder> dict_decoder);
```

Comment:
using BinaryDictPageDecoder's constructor to set dict page decoder means first need to decode slice data header to check whether current page is dict encoding or not,so there is two ways to do it:
1. decode slice data header in FileColumnIterator and getting the encode type,construct dict page decoder and set it to PageDecoderOptions;It works but "decode slice data header" will be done twice
2. decode slice data header and settting the encode type PageDecoderOptions, BinaryDictPageDecoder's no longer to decode the slice data header;but this will change the logic of BinaryDictPageDecoder->init

so I think first call BinaryDictPageDecoder->init to get encoding type and then set dict page decoder to BinaryDictPageDecoder is better

Reply:
I see. Passing dict page decoder via constructor will force us to always read dict page even when only non-dict-encoded data pages are seeked.

I agree with you now to lazy set dict decoder via member function. Two more suggestions

1. remove `dict_decoder` from PageDecoderOptions
2. the argument type should be `PageDecoder*`, since dict decoder is owned by FileColumnIterator

Reply:
👌